compiler/grainlsp/rpc.re

```diff
@@ -0,0 +1,424 @@
+    | Some(id) => Message(id, action, json)
+    };
+  } else {
+    failwith("Invalid header");
+  };
+};
+
+let send = (output, content) => {
+  let length = String.length(content);
+  let sep = Sys.os_type == "Unix" ? "\r\n\r\n" : "\n\n";
+  let len = string_of_int(length);
+
+  output_string(output, "Content-Length: " ++ len ++ sep ++ content);
+  flush(output);
+};
+
+let send_null_message = (output, id) => {
+  let res =
+    `Assoc([
+      ("jsonrpc", `String("2.0")),
+      ("id", `Int(id)),
+      ("result", `Null),
+    ]);
+
+  let strJson = Yojson.Safe.pretty_to_string(res);
+  send(output, strJson);
+};
+
+let send_capabilities = (output, id: int) => {
+  let completion_vals: completion_values = {
+    resolveProvider: true,
+    triggerCharacters: ["."],
+  };
+
+  let codeVals: code_values = {resolveProvider: true};
+
+  let capabilities: lsp_capabilities = {
+    documentFormattingProvider: false,
+    textDocumentSync: 1,
+    hoverProvider: true,
+    completionProvider: completion_vals,
+    definitionProvider: false, // disabled until we can resolve the external module location
+    typeDefinitionProvider: false,
+    referencesProvider: false,
+    documentSymbolProvider: false,
+    codeActionProvider: false,
+    codeLensProvider: codeVals,
+    documentHighlightProvider: false,
+    documentRangeFormattingProvider: false,
+    renameProvider: false,
+  };
+
+  let response: capabilities_response = {
+    jsonrpc,
+    id,
+    result: {
+      capabilities: capabilities,
+    },
+  };
+
+  let res = capabilities_response_to_yojson(response);
+  let strJson = Yojson.Safe.to_string(res);
+
+  send(output, strJson);
+};
+
+let send_lenses = (~output, ~id: int, lenses: list(lens_t)) => {
+  let convertedLenses =
+    List.map(
+      (l: lens_t) => {
+        let rstart: position = {line: l.line - 1, character: 1};
+        let rend: position = {line: l.line - 1, character: 1};
+        let range = {
+          {start: rstart, range_end: rend};
+        };
+
+        let command = {title: l.signature, command: "command string"};
+        let lsp_lens: lsp_lens_t = {range, command};
+        lsp_lens;
+      },
+      lenses,
+    );
+
+  let response: lens_response = {jsonrpc, id, result: convertedLenses};
+  let res = lens_response_to_yojson(response);
+  let strJson = Yojson.Safe.pretty_to_string(res);
+  send(output, strJson);
+};
+
+let send_hover = (~output, ~id: int, ~range: range_t, signature) => {
+  let range_ext = convert_range(range);
+  let hover_info: hover_result = {
+    contents: {
+      kind: "markdown",
+      value: signature,
+    },
+    range: range_ext,
+  };
+  let response: hover_response = {jsonrpc, id, result: hover_info};
+  let res = hover_response_to_yojson(response);
+  let strJson = Yojson.Safe.pretty_to_string(res);
+  send(output, strJson);
+};
+
+let send_diagnostics =
+    (
+      ~output,
+      ~uri,
+      ~warnings: option(list(lsp_warning)),
+      error: option(lsp_error),
```

Comment:
It's weird for error to be the only positional argument here, maybe warnings and error should both be? Unsure